### PR TITLE
Migrated to copier

### PR DESCRIPTION
**Pull request by santacodes**
This is a follow-up on PR #29, the template has been completely migrated to `copier` as [copier](https://github.com/copier-org/copier) is under active development and is better maintained than `cookiecutter`.

While migrating to `copier` I faced some issues and thought I would mention them here, firstly in the `copier.yml` (default configuration file) we cannot add our variables or constants and set their values like we did for `cookiecutter` for `__project_slug` and `__year`. To tackle this problem, I added a prompt for `project_slug` and replaced the `__year` instances with time labels derived from `jinja2_time.TimeExtension`. 
As of now the generated project using `copier` works well and the tests within the generated project are all set, though I have yet to add the template generation tests. For tests I would be using `pytest-copie` linked [here](https://github.com/12rambau/pytest-copie/) which is almost the same as `pytest-cookies` that we are currently using.

## Usage
To generate a project in a directory called `test/` you would need to perform - 
```bash 
copier copy pybamm-cookiecutter/ test/   --trust # This will generate the src layout of the project inside the test directory
# copier prompts the user for details (the --trust flag is to bypass a warning caused by the `jinja2_time.TimeExtension`)
```

### Additional Notes
I will be working on these template generation test cases with all the `build-backends` and also the `cli`. The CLI would wrap everything and just a single command would set up the working directory with project files.

Sub-task #26

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Thanks for taking the initiative to put up this PR early! I think it will be easier to review once we merge #29 and when the diff gets cleaner (and once you modify the template generation tests using `pytest-copie`, as mentioned).

**Comment by agriyakhetarpal** [comment]
Thanks for taking the initiative to put up this PR early! I think it will be easier to review once we merge #29 so that the diff gets cleaner (and once you add the template generation tests using `pytest-copie`, as you mentioned).

File: template/{% if ci == 'gitlab' %}.gitlab-ci.yml{% endif %}.jinja
Comment (brosaplanella) [reply template/{% if ci == 'gitlab' %}.gitlab-ci.yml{% endif %}.jinja:1]: I have never used Gitlab before. If we support it but we haven't fully tested we should say it clearly in the docs or README so users know they use it under their own responsibility.

File: copier.yml
Comment (santacodes) [reply copier.yml:67]: Is it supposed to be called `hatching` or `hatchling`, I am guessing you made a typo in line:66

File: pyproject.toml
Comment (santacodes) [reply pyproject.toml:33]: Yes, that seems fair enough to me, I am curious what others have to say on this.

File: template/pyproject.toml.jinja
Comment (santacodes) [reply template/pyproject.toml.jinja:47]: Maybe we could do something like a script that would go under `_tasks:` key in `copier.yml` which would get the classifiers from the `pybamm` package i.e. `pyproject.toml` of PyBaMM and support those same versions? I am sure there's an easier way to this but that's one of the possible ways I could see right now.

File: template/pyproject.toml.jinja
Comment (santacodes) [reply template/pyproject.toml.jinja:99]: Oops will fix that!

File: template/{% if ci == 'gitlab' %}.gitlab-ci.yml{% endif %}.jinja
Comment (agriyakhetarpal) [reply template/{% if ci == 'gitlab' %}.gitlab-ci.yml{% endif %}.jinja:1]: Sounds like a good idea, since we can't claim compatibility with all CI providers. We can fill out the file by customising the one from `scientific-python/cookie` if we want to (there's no need for attribution since it's just a workflow file).

File: copier.yml
Comment (agriyakhetarpal) [reply copier.yml]: Oops, yes, I did. It is supposed to be `hatchling` with an `l`.

File: pyproject.toml
Comment (brosaplanella) [reply pyproject.toml:33]: If I understand correctly, pybamm is only a dependency for the template, right? So it should not be needed as a dependecy on the project.

File: template/pyproject.toml.jinja
Comment (santacodes) [reply template/pyproject.toml.jinja]: Or maybe we prompt the user for a classifier? that would put copier to use I guess?

File: .github/workflows/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/workflows/test_on_push.yml:121]:
```suggestion
        - name: Install copier and generate a project
```

File: copier.yml
Comment (agriyakhetarpal) [inline copier.yml:14]: I have some reservations about these fields, since people not acquainted with packaging their code might not understand the difference between what a "package" and a "project" are. In theory, packages, whether uploaded on PyPI or not, can be interchangeably called projects in the greater sense of the programming world, which adds to the confusion. I propose something like this:

```suggestion
project_name:
  type: str
  default: "pybamm-example-project"
  help: "The name of your project"

project_slug:
  type: str
  help: The name of the directory containing the Python files for your project. This should be in `snake_case`, and this is what you will import it with, such as `import pybamm_example_project`.
  default: "{{ project_name|lower|replace('-', '_')|replace(' ', '_') }}"
  validator: |
    {% if ('-' in project_slug or ' ' in project_slug) %}
    The name of the directory should not contain spaces or hyphens
    {% endif %}

```

File: copier.yml
Comment (agriyakhetarpal) [inline copier.yml:31]:
```suggestion
  help: "The origin URL to your code repository"
```

File: copier.yml
Comment (agriyakhetarpal) [inline copier.yml:50]:
```suggestion
  default: "A battery modelling project that uses the PyBaMM framework"
```

File: copier.yml
Comment (agriyakhetarpal) [inline copier.yml:67]:
```suggestion
  help: "Choose a build backend. hatch (hatchling) is recommended for pure Python projects"
```

File: template/pyproject.toml.jinja
Comment (agriyakhetarpal) [inline template/pyproject.toml.jinja:19]:
```suggestion
  { name = "The PyBaMM Team", email = "pybamm@pybamm.org" },
```

File: template/pyproject.toml.jinja
Comment (agriyakhetarpal) [inline template/pyproject.toml.jinja:24]:
```suggestion
requires-python = ">=3.8"
```

File: template/pyproject.toml.jinja
Comment (agriyakhetarpal) [reply template/pyproject.toml.jinja]: The project generation should be quick – IMO, such a prompt isn't needed

File: template/pyproject.toml.jinja
Comment (agriyakhetarpal) [inline template/pyproject.toml.jinja:43]:
```suggestion
  "Programming Language :: Python :: 3.11",
```

File: tests/template_tests/test_project_generation.py
Comment (agriyakhetarpal) [inline tests/template_tests/test_project_generation.py:12]: We can also add `"-Lcopie"` to the `codespell` arguments so that we don't need repeated comments. Either way is fine to me.

**Review by agriyakhetarpal — COMMENTED** [review]
Thanks, @santacodes! Could you please fix the conflicts? Happy to approve once the conflicts and the previous conversations are addressed.

File: pyproject.toml
Comment (santacodes) [reply pyproject.toml:33]: I think in order for the entry points to work we would need `pybamm` as a dependency inside the template as the `BaseModel` for all the models is derived from `pybamm`.

**Review by brosaplanella — APPROVED** [review]
Looks good to me, thanks!

File: README.md
Comment (agriyakhetarpal) [inline README.md:38]:
```suggestion
Install `copier` and `jinja2_time` extension using `pip`.
```

File: README.md
Comment (agriyakhetarpal) [inline README.md:43]:
```suggestion
Generate a project from the `pybamm-cookiecutter` template.
**Note:** This requires an internet connection. You could manually clone the git repository and run the copy command, or just execute the copy command with the URL to the git repository.

```

File: README.md
Comment (agriyakhetarpal) [inline README.md:52]:
```suggestion
After generation, you can navigate to the generated project and run `nox -s generated-project-tests` to ensure if the project units are working as intended.
```

File: README.md
Comment (agriyakhetarpal) [inline README.md:56]:
```suggestion
This refers to the project used for the development of this template. There are two ways to install this project: either through `nox` or `pip`. `nox` uses `uv pip` or `pip` internally, and in this case, creates a virtual environment for you to activate.
```

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml:33]: In that case, we can drop the entry points from the project when the template is ready, so that `pybamm-cookiecutter` can provide just the proposed CLI.

**Review by agriyakhetarpal — APPROVED** [review]
Looks great now, thanks, @santacodes!

**Review by arjxn-py — APPROVED** [review]
Looks good @santacodes, this one was fast 🚀 
You've tested this locally too, right?

**Comment by santacodes** [comment]
> Looks good @santacodes, this one was fast 🚀 
> You've tested this locally too, right?

Yes I did test it locally.

**Pull request merged**
